### PR TITLE
Added Coverage Coverage Utility to create xml file and print summray.

### PR DESCRIPTION
**Pull request by harshjain2**
## Description
Added CodeCoverageUtility for
1. Generating XML file from .coverage file.
2. Getting CodeCoverage Summary.

This PR also implements TestLogger interface. 

## Related issue
https://github.com/spekt/vstest.coverage/issues/8

### CODE REVIEW

**Comment by codito** [comment]
I don't think we can check-in and redistribute copyrighted binaries into a github repo. Please exclude codecoverage.* binaries. Let's use those from `~\.nuget\packages\Microsoft.TestPlatform` as needed.

File: src/Coverage.TestLogger.Tests/Microsoft.TestPlatform.Extensions.CoverageLogger.Tests.csproj [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    <TargetFrameworks>netcoreapp2.0</TargetFrameworks>
+    <AssemblyName>Microsoft.TestPlatform.Extensions.CoverageLogger.Tests</AssemblyName>
+    <WarningsAsErrors>true</WarningsAsErrors>
+    <EnableCodeAnalysis>true</EnableCodeAnalysis>
+    <MSTestVersion>1.3.2</MSTestVersion>
+    <NETTestSdkPreviousVersion>15.8.0</NETTestSdkPreviousVersion>
+    <MoqVersion>4.9.0</MoqVersion>
+  </PropertyGroup>
+  <ItemGroup>
+    <None Remove="TestData\CodeCoverageExe\CodeCoverage\amd64\covrun64.dll" />
```
Comment (codito) [inline src/Coverage.TestLogger.Tests/Microsoft.TestPlatform.Extensions.CoverageLogger.Tests.csproj]: Remove please

Comment (harshjain2) [reply src/Coverage.TestLogger.Tests/Microsoft.TestPlatform.Extensions.CoverageLogger.Tests.csproj]: Sure.

File: src/Coverage.TestLogger/CodeCoverageUtility.cs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+                    },
+                    EnableRaisingEvents = true
+                };
+
+                vanguardProcess.Exited += this.CodeCoverageExited;
+                vanguardProcess.Start();
+
+                vanguardProcess.WaitForExit();
+            }
+            catch (DllNotFoundException ex)
```
Comment (codito) [inline src/Coverage.TestLogger/CodeCoverageUtility.cs]: If all the `catch` blocks need to throw, is there a reason to have them? Is some logging intended?

File: src/Coverage.TestLogger/CodeCoverageUtility.cs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+            }
+        }
+
+        private void CodeCoverageExited(object sender, EventArgs e)
+        {
+            if (this.vanguardProcess != null)
+            {
+                if (this.vanguardProcess.HasExited && this.vanguardProcess.ExitCode != 0)
+                {
+                    EqtTrace.Error(this.vanguardProcess.StandardError.ReadToEnd());
```
Comment (codito) [inline src/Coverage.TestLogger/CodeCoverageUtility.cs]: May lord of testability forgive our sin :)

File: src/Coverage.TestLogger/CoverageLogger.cs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+                    var summary = this.codeCoverageUtility.GetCoverageSummary(resultFile);
+                    ConsoleOutput.Instance.WriteLine(summary, OutputLevel.Information);
+                }
+            }
+        }
+
+        private string GetCodeCoverageExePath()
+        {
+            return @"C:\Users\harjain\.nuget\packages\microsoft.internal.codecoverage\1.0.9\contentFiles\any\any\CodeCoverage";
```
Comment (codito) [inline src/Coverage.TestLogger/CoverageLogger.cs]: We need to find a way to have this directory on every user's machine :sunglasses:

Comment (codito) [reply src/Coverage.TestLogger/CoverageLogger.cs]: Just a thought, shouldn't we convert this to lcov and display the text from there? We could reuse some code from other library.

File: src/Coverage.TestLogger/CoverageLogger.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+                {
+                    var resultFile = Path.Combine(Path.GetDirectoryName(codeCoverageFile), Path.GetFileNameWithoutExtension(codeCoverageFile) + ".xml");
+                    var arguments = "analyze /output:" + '"' + resultFile + '"' + " " + '"' + codeCoverageFile + '"';
+                    try
+                    {
+                        this.codeCoverageUtility.AnalyzeCoverageFile(arguments, this.GetCodeCoverageExePath());
+                    }
+                    catch (Exception ex)
+                    {
+                        ConsoleOutput.Instance.WriteLine(ex.Message, OutputLevel.Information);
```
Comment (codito) [inline src/Coverage.TestLogger/CoverageLogger.cs]: Is ConsoleOutput coming from TP? How and when did this become public? :godmode:

File: src/Coverage.TestLogger.Tests/Microsoft.TestPlatform.Extensions.CoverageLogger.Tests.csproj [hunk 6]
```diff
@@ -0,0 +1,43 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFrameworks>netcoreapp2.0</TargetFrameworks>
+    <AssemblyName>Microsoft.TestPlatform.Extensions.CoverageLogger.Tests</AssemblyName>
```
Comment (codito) [inline src/Coverage.TestLogger.Tests/Microsoft.TestPlatform.Extensions.CoverageLogger.Tests.csproj]: Should this be spekt.something?

File: src/Coverage.TestLogger/CoverageLogger.cs (hunk 5 above)
Comment (codito) [reply src/Coverage.TestLogger/CoverageLogger.cs]: Let's not take this dependency if possible. It shouldn't be a publicly documented api somehow.

File: src/Coverage.TestLogger.Tests/CodeCoverageUtilityTests.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+        [TestInitialize]
+        public void Initialize()
+        {
+            this.codeCoverageUtility = new CodeCoverageUtility();
+            var testContainerDirectory = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+            this.sampleCodeCoverageFilePath = Path.Combine(testContainerDirectory, @"TestData\Sample.coverage");
+            this.expectedXMLFilePath = Path.Combine(testContainerDirectory, @"TestData\Expected.xml");
+            this.sampleXMLFilePath = Path.Combine(testContainerDirectory, @"TestData\Sample.xml");
+            this.codeCoverageExePath = Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location), @"..\..\..\..\..\packages\microsoft.codecoverage\15.8.0\build\netstandard1.0\CodeCoverage");
```
[the remark this answers is not in the document]
Comment (harshjain2) [reply src/Coverage.TestLogger.Tests/CodeCoverageUtilityTests.cs]: Added explicit dependency of Microsoft.CodeCoverage nuget package.

File: src/Coverage.TestLogger.Tests/Microsoft.TestPlatform.Extensions.CoverageLogger.Tests.csproj (hunk 6 above)
Comment (harshjain2) [reply src/Coverage.TestLogger.Tests/Microsoft.TestPlatform.Extensions.CoverageLogger.Tests.csproj]: Done.

File: src/Coverage.TestLogger/CodeCoverageUtility.cs (hunk 2 above)
Comment (harshjain2) [reply src/Coverage.TestLogger/CodeCoverageUtility.cs]: Removed.

File: src/Coverage.TestLogger/CoverageLogger.cs (hunk 5 above)
Comment (harshjain2) [reply src/Coverage.TestLogger/CoverageLogger.cs]: Used Console.WriteLine for now.

**Comment by harshjain2** [comment]
There is some scope for improvement which can be done in ConsoleLogger. 
For Hackathon, we will be using MSBuild task and will be passing the Path to CodeCoverage.exe as part of MSBuild arguments. 

Pending workitems that are not covered in this PR:
1. Tests for CoverageLogger.
2. UnitTests. 
3. Way to figure out the location of `Microsoft.CodeCoverage` NuGet package path. Below link can be used to figure out the location of NuGet package folder. https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Configuration/Utility/SettingsUtility.cs#L139

**Pull request merged**